### DISCUSSION TITLE
Feature Request: Support alias in JOIN condition expression

### DISCUSSION THREAD

**Discussion by tekumara** [duckdb/duckdb#11962] [Ideas]
Support aliases in JOIN condition expression, eg:

Given
```sql
CREATE OR REPLACE TEMPORARY TABLE TEST (COL VARCHAR);
INSERT INTO TEST (COL) VALUES ('VARCHAR1'), ('VARCHAR2');
CREATE OR REPLACE TEMPORARY TABLE JOINED (COL VARCHAR, ANOTHER VARCHAR);
INSERT INTO JOINED (COL, ANOTHER) VALUES ('CHAR1', 'JOIN');
```

Then support the friendly sql:
```sql
SELECT
            T.COL
            , SUBSTR(T.COL, 4) AS ALIAS
            , J.ANOTHER
        FROM TEST AS T
        LEFT JOIN JOINED AS J
        ON ALIAS = J.COL;
```

Which currently errors with
```
Error: Binder Error: Referenced column "ALIAS" not found in FROM clause!
Candidate bindings: "J.COL"
LINE 13:         ON ALIAS = J.COL;
```

This friendly SQL dialect would avoid having to repeat the column, ie: its nicer than
```sql
SELECT
            T.COL
            , SUBSTR(T.COL, 4) AS ALIAS
            , J.ANOTHER
        FROM TEST AS T
        LEFT JOIN JOINED AS J
        ON SUBSTR(T.COL, 4)  = J.COL;
```

And also mimics support in other database (eg: Snowflake)

**Comment by Tishj** [comment]
Not the expert on this topic but:

I think this doesn't work because it's looking for the alias on the from clause, which is just T(...), only the final projection list has the aliases, which it won't start evaluating/binding until the from clause is resolved

**Comment by Tishj** [comment]
```sql
SELECT
            T.COL
            , T.ALIAS
            , J.ANOTHER
        FROM (select COL, SUBSTR(COL, 4) ALIAS FROM TEST) T
        LEFT JOIN JOINED AS J
        ON T.ALIAS = J.COL;
```

This is proven by the rewrite to this making it work
It does take column aliases into account, but it can't use aliases of the final projection as those can't be bound until the from (the join) is resolved
